src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -476,7 +501,15 @@ private void setState(State state) {
   }
 
   private boolean shouldRotate(long now) {
-    boolean periodicRotation = rotateIntervalMs > 0 && now - lastRotate >= rotateIntervalMs;
+    currentTimestamp = timestampExtractor.extract(currentRecord);
+    lastRotate = lastRotate == null && currentRecord != null
```

Comment:
Is this supposed to be updating `lastRotate`? The method name doesn't really indicate it would do so. Perhaps this was supposed to be a local variable?

Reply:
This update to the global `lastRotate` needs to be added for timestamp based partitioners for the following reason: 

The only place that `lastRotate` was updated so far was `updateRotationTimers`. However with timestamp based partitioners the case is that in the first call of `updateRotationTimers` and before entering the while loop with `buffer` non-empty, the timestamp-based partitioners don't have a value for `lastRotate`. Then when they enter the loop, they need a call to both: 
```java 
currentTimestamp = timestampExtractor.extract(currentRecord);
lastRotate = lastRotate == null ? currentTimestamp : lastRotate;
```
somewhere, or they'll never rotate based on time. Draining the `buffer` does not make a difference, because the branch on empty `buffer` still calls `shouldRotate` first before calling `updateRotationTimers`. 

Happy to change the name for `shouldRotate` to reflect this change. The reason we don't do this already is probably that we count on processing a batch of records will always take less time than the rotation interval. That's why `now` is passed as argument and is not called every time withing `shouldRotate`. But for timestamp based partitioners this is not an option. Every record needs to be inspected. 

The newly added test case `testWriteRecordTimeBasedPartitionFieldTimestampHours` catches this case. 
(I'm adding some more test cases for record timestamp, but unfortunately a few changes in storage-common are needed too, or a small duplication of code to add `createRecordWithTimestamp`).